### DISCUSSION TITLE
Rectangle bounds are not right.

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16275] [Q&A]
My rectangle looks like this...
![image](https://github.com/AvaloniaUI/Avalonia/assets/2453707/bc2b0013-7a55-4cd4-a556-fc3b0d8edaff)

```xml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:v="using:AvaloniaMario.Views"
             xmlns:vm="using:AvaloniaMario.ViewModels"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="AvaloniaMario.Views.SectionView"
             x:DataType="vm:SectionViewModel">
    <Design.DataContext>
        <!-- This only sets the DataContext for the previewer in an IDE,
             to set the actual DataContext for runtime, set the DataContext property in code (look at App.axaml.cs) -->
        <vm:SectionViewModel/>
    </Design.DataContext>
    
    <StackPanel>
        <ScrollViewer>
            <Rectangle Bounds="{Binding Bounds}" Fill="Black" Stroke="Magenta" StrokeThickness="1">
                    
            </Rectangle>
        </ScrollViewer>
    </StackPanel>
</UserControl>
```

```cs
using System;
using Avalonia;
using Avalonia.Media;
using CommunityToolkit.Mvvm.ComponentModel;

namespace AvaloniaMario.ViewModels;

public partial class SectionViewModel : ObservableObject
{
    [ObservableProperty] private long _top;
    [ObservableProperty] private long _left;
    [ObservableProperty] private long _right = 256;
    [ObservableProperty] private long _bottom = 240;
    [ObservableProperty] private uint _scale = 1;
    
    public Rect Bounds
    {
        get
        {
            Left   = Math.Clamp(Left  , 0, Right  - 256);
            Top    = Math.Clamp(Top   , 0, Bottom - 240);
            Right  = Math.Clamp(Right , Left  + 256, long.MaxValue);
            Bottom = Math.Clamp(Bottom, Top   + 240, long.MaxValue);
            
            return new Rect(Left * Scale, Top * Scale, (Right - Left) * Scale, (Bottom - Top) * Scale);
        }
    }
}
```

**Comment by maxkatz6** [comment]
1. You probably need to set UseLayoutRounding=false on canvas or scroll viewer.
2. ScrollViewer inside of StackPanel is a weird choice - stack panel does not restrain vertical space (or horizontal, if Orientation=Horizontal), so it scroll viewer won't be limited in space, and never will show the scroll bar.

I also recommend joining our telegram group for quicker questions https://t.me/Avalonia.

**Comment by stevemonaco** [comment]
I don't think `Bounds` is intended to be directly set, especially via bindings. I'm getting the following binding error on 11.1.0-rc2:

```
[Binding]An error occurred binding 'Bounds' to 'Bounds' at 'Bounds': 'The property Bounds is readonly.' (Rectangle #7715150)
```

It's defined as a `protected set`:
https://github.com/AvaloniaUI/Avalonia/blob/56108c1caf51122ca6cdf462b5df2b53cf6abae8/src/Avalonia.Base/Visual.cs#L171-L175

You should rethink your layout design beyond the `StackPanel` being a parent for a `ScrollViewer`. `Canvas` is made for absolute positioning within its layout space and is probably a better fit if you want that type of positioning. If `Rectangle` is intended be a tile (judging by your previous issues), then you're probably going to have a bad time with thousands of tiles in the visual tree, as I mentioned before. Though you can fairly quickly test ballpark performance of that using buttons (which are still heavier than a `Rectangle`). Tile maps typically have a dense representation that don't necessarily require a `Canvas`, at least until any sprite layers.

**Reply by maxkatz6** [reply]
Oh, I haven't noticed that. You are right.

**Reply by Shadowblitz16** [reply]
I tried to simplify my layout and moved the scroll view to a outer class...

But it seems rectangle doesn't take Auto or * parameters for with and height either.
I am trying to make the rectangle take up the whole canvas size.

```xml

    <Canvas Background="Black" Left="{Binding Left}" Top="{Binding Top}" Right="{Binding Right}" Bottom="{Binding Bottom}">
        <Rectangle Stroke="Magenta" StrokeThickness="3" Width="*" Height="*"/>
    </Canvas>
```

infact rectangle seems to break it.
![image](https://github.com/AvaloniaUI/Avalonia/assets/2453707/5b7ccac4-89ec-476e-8fea-beab3de57b60)

**Reply by Shadowblitz16** [reply]
Also binding to the size makes it too small...
```xml
    <Canvas Background="Black" Left="{Binding Left}" Top="{Binding Top}" Right="{Binding Right}" Bottom="{Binding Bottom}">
        <Rectangle Stroke="Magenta" StrokeThickness="3" Width="{Binding Right}" Height="{Binding Bottom}"/>
    </Canvas>
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/2453707/cd319beb-9551-43e5-bb65-fd9b664ca843)

**Reply by stevemonaco** [reply]
Auto and `*` are only valid within the context of a `GridLength` which is only used for `Grid`. They're not part of the general layout system.

I'm not sure why you think those dimensions are "too small"? You're explicitly setting them in the VM. 256x240 px is pretty small on a modern desktop although there may be some mistakes. The clamping logic seemed quite weird when I looked at it.

**Reply by Shadowblitz16** [reply]
> Auto and `*` are only valid within the context of a `GridLength` which is only used for `Grid`. They're not part of the general layout system.
> 
> I'm not sure why you think those dimensions are "too small"? You're explicitly setting them in the VM. 256x240 px is pretty small on a modern desktop although there may be some mistakes. The clamping logic seemed quite weird when I looked at it.

yes but the canvas and rectangle use the same size why is the rectangle smaller?

**Reply by stevemonaco** [reply]
Left/Right/Top/Bottom are attached properties: they're intended to be set on direct children of the Canvas so that the Canvas can position them during layout (ie. `<Rectangle Canvas.Left="{Binding Left}" ... />`. They have no effect on the Canvas itself. I'm not really sure why they're directly accessible without the `Canvas.` prefix to indicate static access of an attached property.

**Reply by stevemonaco** [reply]
I suggest you learn how to use DevTools to view the layout while debugging. It won't provide all the answers, but it will let you better view certain things, especially as you get practice. In particular, you should see the `Canvas`'s `HorizontalAlignment` and `VerticalAlignment` to be `Stretch`. Then you can see what the parent layout control's bounds are, likely the entire window.

**Reply by Shadowblitz16** [reply]
> Left/Right/Top/Bottom are attached properties: they're intended to be set on direct children of the Canvas so that the Canvas can position them during layout (ie. `<Rectangle Canvas.Left="{Binding Left}" ... />`. They have no effect on the Canvas itself. I'm not really sure why they're directly accessible without the `Canvas.` prefix to indicate static access of an attached property.

I'm not saying Its having effect on the canvas, I'm saying the canvas has bounds of x,y,256,240 and the rectangle has bounds of 0,0,256,240 the rectangle should take up the whole space.

**Reply by stevemonaco** [reply]
The Canvas doesn't have those bounds. Open up DevTools and inspect it. Top/Left/Right/Bottom aren't used. Only stretching and MinHeight, Height, MaxHeight, and corresponding widths are.

Even if the Canvas were 256x240 px and the Rectangle inside were 256x240 px, the rect still wouldn't take up the whole screen. Stretching dimensions doesn't rescale the visual. That's what Viewbox (or `ScaleTransform`) are for.

Otherwise, provide a repro so I can work with it locally.
